### PR TITLE
Keep QM output files

### PR DESCRIPTION
**Pull request by juraskov**
This PR allows copying output files from QM computations and saves them into designated folders.  Data are saved from initial computations, AL cycle and comparisons function.

TODO:

- [x] Save computations from energy selector (or not?)
- [x] Save computations from initial labelling of the training data
- [x] Correct saving computations from compare 
- [x] Add tests

### CODE REVIEW

File: mlptrain/training/active.py [hunk 1]
```diff
@@ -147,6 +148,8 @@ def train(
         box_size: (List | None) Size of the box where MLP-MD propagated.
 
         keep_al_trajs: (bool) If True, MLP-MD trajectories generated during AL phase are saved into new folder.
+
+        keep_output_files: (bool) If True, outputs of QM computations are saved to new folder.
```
Comment (danielhollas) [inline mlptrain/training/active.py]:
```suggestion
        keep_output_files: (bool) If True, output files of QM computations are saved to new folder.
```

File: mlptrain/configurations/configuration_set.py [hunk 2]
```diff
@@ -851,9 +878,12 @@ def _comparison_name(*args):
         return name
 
 
-def _single_point_eval(config, method_name, **kwargs):
+def _single_point_eval(config, method_name, output_name, **kwargs):
     """Top-level hashable function useful for multiprocessing"""
-    config.single_point(method_name, **kwargs)
+    if 'index' in kwargs:
```
Comment (danielhollas) [inline mlptrain/configurations/configuration_set.py]: This feels like an unfortunate hack. How much work would it be to change the callers of this function to construct the proper `output_name` before calling this function?

File: tests/test_configuration.py [hunk 3]
```diff
@@ -8,6 +10,10 @@
 import numpy as np
 import random
 import pytest
+import os
+from mlptrain.potentials._base import MLPotential
+
+ade.config._ConfigClass.max_core = ade.values.Allocation(1, units='GB')
```
Comment (danielhollas) [inline tests/test_configuration.py]: Thanks for fixing this! Might be nice as a separate PR, but I don't care much.

File: tests/test_configuration.py
Comment (danielhollas) [inline tests/test_configuration.py]: Nice tests, thanks! We might need to test some more edge cases (e.g. what happens if file already exists in QM_Outputs etc). Could also be a follow-up.

**Review by danielhollas — CHANGES_REQUESTED** [review]
This is really impressive work! I have just one major concern with regards to how files are moved to the `QM_outputs` but otherwise this broadly looks good and is better to merge and iterate upon if needed in subsequent PRs.

File: mlptrain/configurations/configuration.py [hunk 4]
```diff
@@ -699,11 +703,49 @@ def single_point(
...
+            )
+
+            if keep_output_files:
+                if output_name is None:
+                    output_name = method
+                    shutil.move(
+                        src=f'tmp_{method}{kept_substrings_list[0]}',
+                        dst=f'QM_outputs/{method}{kept_substrings_list[0]}',
+                    )
+                elif 'energy' in output_name:
```
[the remark this answers is not in the document]
Comment (juraskov) [reply mlptrain/configurations/configuration.py:755]: These are data generated by the energy selector - most of them will not be very relevant, because if the MLIP works well for the data point, the structure is not added to the data set. At the same time, this can be potentially a lot of data, which I decided to discard.

File: mlptrain/configurations/configuration_set.py (hunk 2 above)
Comment (juraskov) [reply mlptrain/configurations/configuration_set.py:889]: I don't understand what you mean here. Happy to discuss in person.

File: tests/test_configuration.py [hunk 5]
```diff
@@ -328,3 +334,85 @@ def test_mol_dict_corrupt_file(tmp_path):
...
+
+            else:
+                raise ValueError(
+                    'Cannot predict the energy and forces on ' f'{type(arg)}'
+                )
+
+        for configuration in all_configurations:
+            # evaluate predicted energies and forces
+            configuration.energy.predicted = 10.0
+            configuration.forces.predicted = [0.0, 0.0, 0.0]
```
[the remark this answers is not in the document]
Comment (juraskov) [reply tests/test_configuration.py:387]: Yes, it should, but we currently test on water only.

File: mlptrain/configurations/configuration.py [hunk 6]
```diff
@@ -713,11 +719,47 @@ def single_point(
...
+            keep_output_files: If true, copy back the QM outputs.
         """
         implemented_methods = ['xtb', 'orca', 'g09', 'g16']
 
+        if keep_output_files:
+            os.makedirs('QM_outputs', exist_ok=True)
+
         if isinstance(method, str) and method.lower() in implemented_methods:
-            run_autode(self, method, n_cores=n_cores)
+            if keep_output_files:
```
Comment (juraskov) [inline mlptrain/configurations/configuration.py:731]:
```suggestion
            if keep_output_files:
                kept_substrings_list = []
```

**Review by juraskov — COMMENTED** [review]
Thanks for the comments, I committed the suggested changes and responded to the comments.

File: mlptrain/configurations/configuration.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (danielhollas) [reply mlptrain/configurations/configuration.py:755]: I see, makes sense. Let's discuss in person. At the very least there should be a comment here explaining what you just wrote.

File: mlptrain/training/selection.py [hunk 7]
```diff
@@ -102,9 +102,11 @@ def __call__(self, configuration, mlp, **kwargs) -> None:
 
         if configuration.energy.predicted is None:
             self._configuration.single_point(mlp)
-
         self._configuration.single_point(
-            method_name, n_cores=kwargs.get('n_cores', 1)
+            method_name,
+            n_cores=kwargs['n_cores'],
+            keep_output_files=kwargs['keep_output_files'],
```
Comment (danielhollas) [inline mlptrain/training/selection.py:108]: @juraskov why not set `keep_output_files` to False here in the energy selector? Then we wouldn't need to special case the `single_point` method

File: tests/test_configuration.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (danielhollas) [reply tests/test_configuration.py:387]: Yes, but shouldn't it be a 3x3 matrix of forces for water?

File: mlptrain/training/selection.py (hunk 7 above)
Comment (juraskov) [reply mlptrain/training/selection.py:108]: But I still would like to save the files which corresponds to the selected structures.

File: mlptrain/configurations/configuration.py [hunk 8]
```diff
@@ -702,6 +704,9 @@ def single_point(
         self,
         method: Union[str, MLPotential],
         n_cores: int = 1,
+        keep_output_files: bool = True,
+        output_name: Optional[str] = None,
```
Comment (danielhollas) [inline mlptrain/configurations/configuration.py]:
```suggestion
        output_name: str | None = None,
```

File: mlptrain/configurations/configuration.py [hunk 9]
```diff
@@ -713,11 +718,48 @@ def single_point(
...
 
             n_cores: Number of cores to use for the calculation
+
+            keep_output_files: If true, copy back the QM outputs.
         """
         implemented_methods = ['xtb', 'orca', 'g09', 'g16']
 
+        if keep_output_files:
+            os.makedirs('QM_outputs', exist_ok=True)
+
```
[the remark this answers is not in the document]
Comment (danielhollas) [reply mlptrain/configurations/configuration.py]: Agree with this, but would implement the change differently, simply moving lines 726-727 below if condition on line 729

File: mlptrain/configurations/configuration.py [hunk 10]
```diff
@@ -713,11 +718,48 @@ def single_point(
...
+                **kwargs,
+            )
+
+            if keep_output_files:
+                if output_name is None:
+                    output_name = method
+                    shutil.move(
+                        src=f'tmp_{method}{kept_substrings_list[0]}',
+                        dst=f'QM_outputs/{method}{kept_substrings_list[0]}',
+                    )
```
[the remark this answers is not in the document]
Comment (danielhollas) [reply mlptrain/configurations/configuration.py]: This probably should be fixed before merging.

**Review by danielhollas — APPROVED** [review]
Given that this PR has been extensively tested in practice, I'd propose to merge it basically as is, and then do any potential improvements / cleanup in a separate PR.
